walmart/walmart.py

```diff
@@ -2,32 +2,65 @@
 import requests
 import uuid
 import base64
-import time
-from uuid import uuid4
 from lxml import etree
 from lxml.builder import E, ElementMaker
-from Crypto.Hash import SHA256
-from Crypto.PublicKey import RSA
-from Crypto.Signature import PKCS1_v1_5
 
-try:
-    from urllib.parse import urlencode
-except ImportError:
-    from urllib import urlencode
+from .exceptions import WalmartAuthenticationError
 
 
 class Walmart(object):
 
-    def __init__(self, consumer_id, channel_type, private_key):
-        self.base_url = 'https://marketplace.walmartapis.com/v2/%s'
-        self.consumer_id = consumer_id
-        self.channel_type = channel_type
-        self.private_key = private_key
-        self.session = requests.Session()
-        self.session.headers['Accept'] = 'application/xml'
-        self.session.headers['WM_SVC.NAME'] = 'Walmart Marketplace'
-        self.session.headers['WM_CONSUMER.ID'] = self.consumer_id
-        self.session.headers['WM_CONSUMER.CHANNEL.TYPE'] = self.channel_type
+    def __init__(self, client_id, client_secret):
+        """To get client_id and client_secret for your Walmart Marketplace
+        visit: https://developer.walmart.com/#/generateKey
+        """
+        self.client_id = client_id
+        self.client_secret = client_secret
+        self.token = None
+        self.token_expires_in = None
+        self.base_url = "https://marketplace.walmartapis.com/v3"
+
+        session = requests.Session()
+        session.headers.update({
+            "WM_SVC.NAME": "Walmart Marketplace",
+            "Content-Type": "application/x-www-form-urlencoded",
+            "Accept": "application/json",
+        })
+        self.session = session
+
+        # Get the token required for API requests
+        self.authenticate()
+
+    def authenticate(self):
+        # Python 3 b64encode expects bytes instead of string to encode,
+        # whereas requests library expects strings in headers. So decode the
+        # bytes generated from b64encode to string
+        encoded_token = base64.b64encode((
+            "{}:{}".format(self.client_id, self.client_secret)
+        ).encode("utf-8")).decode("utf-8")
+
+        self.session.headers["Authorization"] = "Basic {}".format(
+            encoded_token
+        )
+        response = self.send_request(
+            "POST", "{}/token".format(self.base_url),
+            body={
+                "grant_type": "client_credentials",
+            },
+        )
+        try:
+            response.raise_for_status()
+        except requests.exceptions.HTTPError:
+            raise WalmartAuthenticationError((
+                "Invalid client_id or client_secret. Please verify your "
+                "credentials from https://developer.walmart.com/#/generateKey"
+            ))
+        else:
+            data = response.json()
+            self.token = data["access_token"]
+            self.token_expires_in = data["expires_in"]
```

Comment:
are you checking if token is expired before making any request?

Reply:
Just keeping it, if in case the user wants to keep track of it. Although this library automatically refreshes the token